File: cpp/samchon/protocol/master/MediatorSystem.cpp
#include <samchon/protocol/master/MediatorSystem.hpp>

#include <samchon/protocol/external/ExternalClientArray.hpp>
#include <samchon/protocol/InvokeHistory.hpp>

using namespace std;
using namespace samchon;
using namespace samchon::library;
using namespace samchon::protocol;
using namespace samchon::protocol::master;
using namespace samchon::protocol::external;

MediatorSystem::MediatorSystem(ExternalClientArray *systemArray)
{
	this->system_array = systemArray;
}
MediatorSystem::~MediatorSystem()
{
}

void MediatorSystem::replyData(shared_ptr<Invoke> invoke)
{
	system_array->replyData(invoke);
}

void MediatorSystem::notify_end(size_t uid)
{
	if (progress_list.has(uid) == false)
		return;

	shared_ptr<InvokeHistory> history = progress_list.get(uid);
	progress_list.erase(history->getUID());

	this->sendData(history->toInvoke());
}

File: ts/src/samchon/protocol/master/MediatorSystem.ts
/// <reference path="../../API.ts" />

/// <reference path="../slave/SlaveSystem.ts" />

namespace samchon.protocol.master
{
	export abstract class MediatorSystem
		extends slave.SlaveSystem
	{
		private system_array: ParallelSystemArray;
		private progress_list: std.HashMap<number, InvokeHistory>;

		public constructor(systemArray: ParallelSystemArray)
		{
			super();

			this.system_array = systemArray;
			this.progress_list = new std.HashMap<number, InvokeHistory>();
		}

		public abstract start(): void;

		public notifyEnd(uid: number): void
		{
			if (this.progress_list.has(uid) == false)
				return;

			let history: InvokeHistory = this.progress_list.get(uid);
			
			this.sendData(history.toInvoke());

			this.progress_list.erase(uid);
		}
	}
}